hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -2732,3 +2741,228 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
+
+	return nil
+}
+
+func reconcileMachineApproverRole(role *rbacv1.Role) error {
+	role.Rules = []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{"cluster.x-k8s.io"},
+			Resources: []string{"machines", "machines/status"},
+			Verbs:     []string{"get", "list", "watch"},
+		},
+	}
+	return nil
+}
+
+func reconcileMachineApproverRoleBinding(binding *rbacv1.RoleBinding, role *rbacv1.Role, sa *corev1.ServiceAccount) error {
+	binding.RoleRef = rbacv1.RoleRef{
+		APIGroup: "rbac.authorization.k8s.io",
+		Kind:     "Role",
+		Name:     role.Name,
+	}
+
+	binding.Subjects = []rbacv1.Subject{
+		{
+			Kind:      "ServiceAccount",
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		},
+	}
+	return nil
+}
+
+func reconcileMachineApproverDeployment(deployment *appsv1.Deployment, hc *hyperv1.HostedCluster, sa *corev1.ServiceAccount, kubeConfigSecret *corev1.Secret, config *corev1.ConfigMap, machineApproverImage, machineApproverVersion string) error {
+	args := []string{
+		"--config=/var/run/configmaps/config/config.yaml",
+		"-v=4",
+		"--logtostderr",
+		"--apigroup=cluster.x-k8s.io",
+		"--workload-cluster-kubeconfig=/etc/kubernetes/kubeconfig/kubeconfig",
+		"--machine-namespace=" + deployment.Namespace,
+	}
+
+	deployment.Spec = appsv1.DeploymentSpec{
+		Replicas: k8sutilspointer.Int32Ptr(1),
+		Selector: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"app": "machine-approver",
+			},
+		},
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					"app": "machine-approver",
+				},
+				Name: "machine-approver",
+			},
+			Spec: corev1.PodSpec{
+				ServiceAccountName: sa.Name,
+				Tolerations: []corev1.Toleration{
+					{
+						Key:    "node-role.kubernetes.io/master",
+						Effect: corev1.TaintEffectNoSchedule,
+					},
+				},
+				Volumes: []corev1.Volume{
+					{
+						Name: "kubeconfig",
+						VolumeSource: corev1.VolumeSource{
+							Secret: &corev1.SecretVolumeSource{
+								SecretName: kubeConfigSecret.Name,
+							},
+						},
+					},
+					{
+						Name: "config",
+						VolumeSource: corev1.VolumeSource{
+							ConfigMap: &corev1.ConfigMapVolumeSource{
+								LocalObjectReference: corev1.LocalObjectReference{
+									Name: config.Name,
+								},
+								Optional:    k8sutilspointer.BoolPtr(true),
+								DefaultMode: k8sutilspointer.Int32Ptr(440),
+							},
+						},
+					},
+				},
+				Containers: []corev1.Container{
+					{
+						Name:            "machine-approver-controller",
+						Image:           machineApproverImage,
+						ImagePullPolicy: corev1.PullAlways,
+						VolumeMounts: []corev1.VolumeMount{
+							{
+								Name:      "kubeconfig",
+								MountPath: "/etc/kubernetes/kubeconfig",
+							},
+							{
+								Name:      "config",
+								MountPath: "/var/run/configmaps/config",
+							},
+						},
+						Env: []corev1.EnvVar{
+							{
+								Name:  "RELEASE_VERSION",
+								Value: machineApproverVersion,
```

Comment:
what's this environment variable for?

Reply:
this will be showed as ClusterOperator CR's version. (`oc get co`)

Reply:
I don't think we need to set this nor break the version down from the image name. We are not even deploying the clusterOperator for the machine approver right? The version here is dictated by the Hypershift release.

Reply:
good question, let me try again without setting this env. machine-approver was writing the clusteroperator's status with a version, before I used my customized image, and it was empty, let me try again 👍

Reply:
>machine-approver was writing the clusteroperator's status with a version

The clusterOperator is a cluster scope resource so we can not deploy it here by design. This PR is not even making that possible because is not given clusterOperator rbac which is fine. My expectation would be for the status controller to just being unable to operate because the lack of rbac.

Reply:
oh, for hosted cluster I gave the `service-network-admin-kubeconfig`, so machine-approver has admin permission, and it will be able to update clusteroperator CR on hosted cluster

Reply:
>oh, for hosted cluster I gave the service-network-admin-kubeconfig

can you please point me to where in this PR is that happening?

Reply:
yes, see https://github.com/openshift/hypershift/pull/477#discussion_r716848971

Reply:
Tried upstream 4.10.0 image without the RELEASE_VERSION set, and it will show like the following:
```
# oc get co
NAME                                       VERSION      AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
kube-apiserver                             4.9.0-rc.2   True        False         False      2m16s   
kube-controller-manager                    4.9.0-rc.2   True        False         False      2m16s   
...
machine-approver                           unknown      True        False         False      2m50s  
```

Reply:
># oc get co

Is this in the guest cluster?

Reply:
So the fact that the clusteroperator is created in the guest cluster and that csr and nodes details might be kube version specific would make me reconsider to run the machine approver from the payload by the HCP instead. Though I'd rather avoid that at least in the short term so we don't need to deal with backporting the machine-approver capi support.

So for now in this PR we have 3 options:
- Leave the clusteroperator version unknown
- Set the clusteroperator version, but it will always show 4.10 even for 4.8 HC.
- Add support in the mnachine approver to not run the cluster operator status controller, so it does not get deployed.

My preference would be the first one.
cc @csrwng, @ironcladlou @sjenning

Reply:
yes, `oc get co` is inside guest cluster. 

I updated to use the first option, and it will show version=unknown for now 👍

Reply:
@enxebre for now I think it's fine to go with option 1, although if possible I'd like to eventually implement option 3. There is no part of the approver that is running in the guest cluster.